proto/stream_service.proto

```diff
@@ -108,10 +108,14 @@ message StreamingControlStreamRequest {
 
 message StreamingControlStreamResponse {
   message InitResponse {}
+  message ShutdownResponse {
+    bool has_running_actor = 1;
+  }
 
   oneof response {
     InitResponse init = 1;
     BarrierCompleteResponse complete_barrier = 2;
+    ShutdownResponse shutdown = 3;
```

Comment:
I am thinking that, is this `ShutdownResponse` necessary? The main difference between receiving an explicit shutdown signal and receiving an disconnected err from the stream is that, in this shutdown response we can know whether there are remaining running actors from the response.

However, this information seems unnecessary, because we can know whether we need to collect barrier from this node from the epoch queue in the `CheckpointControl`, and we can decide whether to trigger recovery by checking if we have any pending epoch waiting on barrier collection from this node.

Reply:
> However, this information seems unnecessary, because we can know...

This is true. However, I still introduced this new variant of responses even when it's empty at the beginning of this PR. 

> The main difference between receiving an explicit shutdown signal and receiving an disconnected err from the stream

I suppose there are two alternatives:

- Explicitly send an `Err(Shutdown)` through the stream from the compute node.
  We need to be able to recognize this error and differentiate it from other errors in the meta node, for (slightly) different handling behavior, e.g., do not ignore this error even if it's not associated with an in-flight barrier.

- Close the channel and let meta service acknowledge it by receiving a `None`.
  I'm just concerned whether this is reliable enough, i.e., if there are any other unexpected scenarios where the stream is disconnected without an error.

Reply:
> for (slightly) different handling behavior, e.g., do not ignore this error even if it's not associated with an in-flight barrier

I think the behavior can be unified when we receive any error from the control stream and know that it's not associated with an inflight-barrier. For example, when a compute node with no inflight barrier gets killed by accidence, even if it's not gracefully shutdown, we should not trigger a recovery. The complexity can be greatly reduced if we just unify the way we handle all kinds of error from the control stream.

Reply:
> It's also possible that the next periodic barrier hasn't been scheduled yet.

If so, when handling the next barrier, the inject_barrier will fail because there will be actors to collect on this node, but the node has been removed and cannot get the control stream, and then it will trigger a recovery.

Reply:
I just find it involve too much indirection. Users may sometimes see messages like "the node is shutting down" or sometimes "unconnected worker node". Since we're sure what is happening in this case, a separate path to simplify the control stream seems to be less confusing, at least for me.

Reply:
Having a `Shutdown` enum variant to explicitly mark the end of control stream LGTM, but the `has_running_actor` seems unnecessary. It's not a ground truth information of whether there is running actors on this CN.

In current PR, we use `self.actor_manager_state.handles.is_empty()` as the value of `has_running_actor`. However, this information is not correct. It may mix with stopped actors and actors that just gets spawned but not join the streaming graph yet.

We can get the ground truth information on meta node. We can check the epoch node queue to see if there is any epoch waiting on barrier collection on the node. For case that all epochs have completed, we can further check the `InflightActorInfo` to see whether there is running actor to send barrier on this node. If there is no, we can ignore the shutdown and do not trigger recovery.

Reply:
> but the `has_running_actor` seems unnecessary.

You're right. Considering that there will be some changes in following refactoring, I would prefer leaving this optimization in future PRs. 🥺

Reply:
Sure. You can merge this PR first and in https://github.com/risingwavelabs/risingwave/pull/17758 I have included this optimization.